quickwit-janitor/src/retention_policy_evaluation.rs

```diff
@@ -0,0 +1,98 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use quickwit_actors::ActorContext;
+use quickwit_config::{RetentionPolicy, RetentionPolicyCutoffReference};
+use quickwit_metastore::{Metastore, Split, SplitMetadata, SplitState};
+use time::OffsetDateTime;
+
+use crate::actors::RetentionPolicyEvaluator;
+
+/// Detect all discardable splits based a retention policy and mark then as MarkedForDeletion.
+///
+/// * `index_id` - The target index id.
+/// * `metastore` - The metastore managing the target index.
+/// * `retention_policy` - The retention policy to used to evaluate the splits.
+/// * `ctx_opt` - A context for reporting progress (only useful within quickwit actor).
+pub async fn run_evaluate_retention_policy(
+    index_id: &str,
+    metastore: Arc<dyn Metastore>,
+    retention_policy: &RetentionPolicy,
+    ctx_opt: Option<&ActorContext<RetentionPolicyEvaluator>>,
+) -> anyhow::Result<Vec<SplitMetadata>> {
+    // Select published splits and filter for eligibility.
+    let current_date_time = OffsetDateTime::now_utc();
+    let discardable_splits: Vec<SplitMetadata> = metastore
+        .list_splits(index_id, SplitState::Published, None, None)
+        .await?
+        .into_iter()
+        // TODO: possibly move this in DB query
+        .filter(|split| is_discardable(current_date_time, split, retention_policy).unwrap_or(false))
+        .map(|meta| meta.split_metadata)
+        .collect();
+    if let Some(ctx) = ctx_opt {
+        ctx.record_progress();
+    }
+
+    if discardable_splits.is_empty() {
+        return Ok(discardable_splits);
+    }
+
+    // Schedule all discardable splits for delete
+    let split_ids: Vec<&str> = discardable_splits
+        .iter()
+        .map(|meta| meta.split_id())
+        .collect();
+    metastore
+        .mark_splits_for_deletion(index_id, &split_ids)
+        .await?;
+
+    Ok(discardable_splits)
+}
+
+/// Checks to see if a split can be discarded based on retention policy.
+fn is_discardable(
+    current_date_time: OffsetDateTime,
+    split: &Split,
+    retention_policy: &RetentionPolicy,
+) -> anyhow::Result<bool> {
+    let retention_period = retention_policy.retention_period()?;
+    let publish_timestamp = split
+        .publish_timestamp
+        .context("Expected a split with a valid published timestamp.")?;
```

Comment:
That's not backward compatible. Use the `update_timestamp` instead for splits produced pre-Quickwit 0.4. Comment the code abundantly.